### PR TITLE
vconsole: avoid dummy console and do more sanity checks before trying to copy fonts

### PR DESCRIPTION
**Pull request by msoltyspl**
See https://github.com/systemd/systemd/issues/4344

### CODE REVIEW

File: src/vconsole/vconsole-setup.c [hunk 1]
```diff
@@ -225,7 +225,7 @@ static void setup_remaining_vcs(int fd, bool utf8) {
                 return;
         }
 
-        fontbuf = malloc(cfo.width * cfo.height * cfo.charcount / 8);
+        fontbuf = malloc(32 / 8 * 32 * 512);
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: i think a comment explaining where these pretty much random numbers come from would be appropriate here...

File: src/vconsole/90-vconsole.rules.in [hunk 2]
```diff
@@ -7,4 +7,4 @@
 
 # Each vtcon keeps its own state of fonts.
 #
-ACTION=="add", SUBSYSTEM=="vtconsole", KERNEL=="vtcon*", RUN+="@rootlibexecdir@/systemd-vconsole-setup"
+ACTION=="add", SUBSYSTEM=="vtconsole", KERNEL=="vtcon*", ATTR{name}!="*dummy device", RUN+="@rootlibexecdir@/systemd-vconsole-setup"
```
Comment (keszybz) [inline src/vconsole/90-vconsole.rules.in:10]: Nice!

File: src/vconsole/vconsole-setup.c (hunk 1 above)
Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c]: Sure

**Comment by keszybz** [comment]
Also please rename the PR to something more meaningful.

File: src/vconsole/vconsole-setup.c [hunk 3]
```diff
@@ -208,9 +208,9 @@ static int font_load_and_wait(const char *vc, const char *font, const char *map,
  */
 static void setup_remaining_vcs(int fd, bool utf8) {
         struct console_font_op cfo = {
-                .op = KD_FONT_OP_GET, .flags = 0,
-                .width = 32, .height = 32,
-                .charcount = 512,
+                .op = KD_FONT_OP_GET,
+                .width = UINT_MAX, .height = UINT_MAX,
+                .charcount = UINT_MAX,
```
Comment (yuwata) [inline src/vconsole/vconsole-setup.c:213]: Why do you remove `.flags=0`? Also, is `.data=NULL` not necessary?

Comment (msoltyspl) [reply src/vconsole/vconsole-setup.c:213]: Yea, if any field is explicitly initialized, any other field gets 0 / null / etc.

Comment (yuwata) [reply src/vconsole/vconsole-setup.c:213]: I see. Thanks.

File: src/vconsole/90-vconsole.rules.in (hunk 2 above)
Comment (poettering) [inline src/vconsole/90-vconsole.rules.in:10]: hmm, so this is racy: by the time the process is started the dummy device might already have disappeared again from the kernel. The above is hence merely an opimization, but doesn't fix anything, does it?

**Comment by yuwata** [comment]
One of my machines which often outputs the `KD_FONT_OP_GET` error messages now works fine. Before these patches, if I remember correctly, systemd-vconsole-setup is called three times during the boot process. Now it called twice. I guess one removed call is for the dummy console.
The following journal log is given by f70ebf1 with this PR.
https://gist.github.com/yuwata/0c3c8ec354dd09c8067b65c2ddc2f5d1

**Comment by keszybz** [comment]
If it disappears, that's a separate issue, and code handles it like before
(EXIT_FAILURE). The fix, as I understand it, is that this way we don't try to
run vconsole-setup on dummy devices.

Or do you mean that a real device could appear, disappear, and a dummy device
could appear with the same name? That's possible, but probably not as likely
as the other way around (i.e. dummy appearing and disappearing rapidly).

**Pull request merged**